esp32/mods/modhttp.c

```diff
@@ -0,0 +1,1475 @@
+
+				return;
+			}
+
+			// Mark the current element as previous, needed when removing the actual current element from the list
+			previous = current;
+		}
+	}
+}
+
+// Update the value of a resource
+STATIC void resource_update_value(mod_http_resource_obj_t* resource, mp_obj_t new_value) {
+
+	// Invalidate current data first
+	resource->value_len = 0;
+	m_free(resource->value);
+
+	if (mp_obj_is_integer(new_value)) {
+
+		uint32_t value = mp_obj_get_int_truncated(new_value);
+		if (value > 0xFF) {
+			resource->value_len = 2;
+		} else if (value > 0xFFFF) {
+			resource->value_len = 4;
+		} else {
+			resource->value_len = 1;
+		}
+
+		// Allocate memory for the new data
+		resource->value = m_malloc(resource->value_len);
+		memcpy(resource->value, &value, sizeof(value));
+
+	} else {
+
+		mp_buffer_info_t value_bufinfo;
+		mp_get_buffer_raise(new_value, &value_bufinfo, MP_BUFFER_READ);
+		resource->value_len = value_bufinfo.len;
+
+		// Allocate memory for the new data
+		resource->value = m_malloc(resource->value_len);
+		memcpy(resource->value, value_bufinfo.buf, resource->value_len);
+	}
+}
+
+
+STATIC esp_err_t mod_http_resource_callback_helper(mod_http_resource_obj_t* resource , httpd_method_t method, mp_obj_t callback){
+
+	esp_err_t ret = ESP_OK;
+
+	/* This needs to be static otherwise, most probably due to compiler optimizations, the
+	 * value of it is not updated between to subsequent calls and httpd_register_uri_handler()
+	 * fails with error ESP_ERR_HTTPD_HANDLER_EXISTS because it gets the URI of the previous resource
+	 */
+	STATIC httpd_uri_t uri;
+
+	// Set the URI
+	uri.uri = resource->uri;
+
+	// Save the user's callback into user context field for future usage
+	uri.user_ctx = callback;
+	// The registered handler is our own handler which will handle different requests and call user's callback, if any
+	uri.handler = mod_http_server_callback;
+
+	// Unregister first URI, solves ESP_ERR_HTTPD_ALLOC_MEM issue
+	ret = httpd_unregister_uri(server_obj->server, uri.uri);
+
+	if((method & MOD_HTTP_GET) && (ret == ESP_OK)) {
+		uri.method = HTTP_GET;
+		ret = httpd_register_uri_handler(server_obj->server, &uri);
+	}
+
+	if((method & MOD_HTTP_PUT) && (ret == ESP_OK)) {
+		uri.method = HTTP_PUT;
+		ret = httpd_register_uri_handler(server_obj->server, &uri);
+	}
+
+	if((method & MOD_HTTP_POST) && (ret == ESP_OK)) {
+		uri.method = HTTP_POST;
+		ret = httpd_register_uri_handler(server_obj->server, &uri);
+	}
+
+	if((method & MOD_HTTP_DELETE) && (ret == ESP_OK)) {
+		uri.method = HTTP_DELETE;
+		ret = httpd_register_uri_handler(server_obj->server, &uri);
+	}
+
+	return ret;
+}
+
+// Compares Accept request-header field with resources mediatype
+STATIC bool mod_http_server_get_acceptance(const char* accept_field, uint8_t mediatype_id) {
+
+	// Start with no acceptance
+	bool accept = false;
+
+	// Initiate resource's mediatype long slice
+	char* slice;
+	slice = (char*)malloc(strlen(mod_http_mediatype[mediatype_id])*(sizeof(char)+1));
```

Comment:
malloc() should not be used. Please use the relevant API whether you want to allocate memory from GC controlled heap (m_malloc()) or from the heap not controlled by the MicroPython's GC: heap_caps_malloc